File: config/plugins/gemini-master-suite/scripts/python/error_checker.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
error_checker.py — Kapsamlı Proje Hata ve Sözdizim Kontrol Aracı (Bug Hunter & QA)

Bu araç belirtilen dizin veya dosyada:
1. PHP Sözdizim Doğrulama (php -l)
2. Python Sözdizim Doğrulama (py_compile)
3. JavaScript / Node.js Sözdizim Doğrulama (node --check)
4. JSON Format Doğrulama (json.loads)
5. UTF-8 BOM Tespiti (\xef\xbb\xbf)
6. Güvenlik ve Mantık Açığı Statik Analizi (eval, XSS, raw SQL)

Kullanım:
    python error_checker.py [hedef_klasor_veya_dosya] [--fix-bom] [--json] [--strict]
"""

import sys
import os
try:
    sys.stdout.reconfigure(encoding="utf-8")
    sys.stderr.reconfigure(encoding="utf-8")
except Exception:
    pass
import re
import json
import subprocess
from datetime import datetime

CANONICAL_PHP = r"php"
CANONICAL_NODE = r"node"
CANONICAL_PYTHON = r"python3"

PHP_EXE = CANONICAL_PHP if os.path.exists(CANONICAL_PHP) else "php"
NODE_EXE = CANONICAL_NODE if os.path.exists(CANONICAL_NODE) else "node"
PYTHON_BIN = 'python3'

IGNORED_DIRS = {
    'node_modules', 'vendor', '.git', 'storage', 'cache',
    'dist', 'build', '.vscode', '.idea', 'tmp', 'temp'
}

BUG_PATTERNS = [
    (re.compile(r"eval\s*\(", re.I), "SECURITY", "SEC-EVAL", "eval() kullanımı tespit edildi"),
    (re.compile(r"\bextract\s*\(\s*\$_(GET|POST|REQUEST)", re.I), "SECURITY", "SEC-EXTRACT", "extract($_GET/$_POST) değişken enjeksiyon riski"),
    (re.compile(r"echo\s+\$_(GET|POST|REQUEST)\[[^\]]+\]\s*;", re.I), "SECURITY", "SEC-XSS", "Doğrudan kullanıcı girdisi echo ediliyor (XSS riski)"),
    (re.compile(r"\$_(GET|POST|REQUEST)\[[^\]]+\]\s*\.\s*(['\"][^'\"]*SELECT|INSERT|UPDATE|DELETE)", re.I), "SECURITY", "SEC-SQLI", "Ham SQL sorgusuna girdi birleştirme"),
    (re.compile(r"\b(shell_exec|exec|system|passthru|popen|proc_open)\s*\(", re.I), "SECURITY", "SEC-CMD-EXEC", "Sistem komut yürütücü çağrısı"),
    (re.compile(r"var_dump\s*\(|console\.log\s*\(", re.I), "QUALITY", "QUAL-DEBUG-LEAK", "Debug çıktısı (var_dump/console.log) kalıntısı")
]

def run_checker():
    target = "."
    fix_bom = False
    is_json = False
    is_strict = False

    args = sys.argv[1:]
    for arg in args:
        if arg == "--fix-bom":
            fix_bom = True
        elif arg == "--json":
            is_json = True
        elif arg == "--strict":
            is_strict = True
        elif not arg.startswith("--"):
            target = arg

    resolved_target = os.path.abspath(target)
    if not os.path.exists(resolved_target):
        print(f"❌ HATA: Hedef bulunamadı: {resolved_target}", file=sys.stderr)
        sys.exit(2)

    report = {
        "target": resolved_target,
        "timestamp": datetime.now().isoformat(),
        "scanned_files": 0,
        "stats": {"pass": 0, "fail": 0, "warn": 0, "info": 0},
        "errors": [],
        "warnings": []
    }

    def add_log(status, code, path_str, msg, detail=""):
        if status == "FAIL":
            report["stats"]["fail"] += 1
            report["errors"].append({"code": code, "file": path_str, "message": msg, "detail": detail})
        elif status == "WARN":
            report["stats"]["warn"] += 1
            report["warnings"].append({"code": code, "file": path_str, "message": msg, "detail": detail})
        elif status == "INFO":
            report["stats"]["info"] += 1
        elif status == "PASS":
            report["stats"]["pass"] += 1

    def scan_file(file_path):
        ext = os.path.splitext(file_path)[1].lower()
        if ext not in {'.php', '.js', '.mjs', '.cjs', '.json', '.py', '.html', '.sql'}:
            return

        report["scanned_files"] += 1
        rel_path = os.path.relpath(file_path, resolved_target)

        try:
            with open(file_path, 'rb') as f:
                raw_bytes = f.read()
        except Exception as e:
            add_log("FAIL", "IO-ERR", rel_path, str(e))
            return

        # BOM kontrolü
        if raw_bytes.startswith(b'\xef\xbb\xbf'):
            if fix_bom:
                with open(file_path, 'wb') as f:
                    f.write(raw_bytes[3:])
                raw_bytes = raw_bytes[3:]
                add_log("WARN", "BOM-FIXED", rel_path, "UTF-8 BOM silindi.")
            else:
                add_log("FAIL", "BOM-DETECTED", rel_path, "UTF-8 BOM tespit edildi.")

        try:
            text_content = raw_bytes.decode('utf-8')
        except UnicodeDecodeError:
            try:
                text_content = raw_bytes.decode('utf-8', errors='replace')
                add_log("WARN", "NON-UTF8", rel_path, "Dosya bozuk karakter içeriyor.")
            except Exception:
                text_content = ""

        # PHP Kontrolü
        if ext == '.php':
            cmd = [PHP_EXE, "-l", file_path]
            try:
                res = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, timeout=10)
                if res.returncode != 0 or "No syntax errors detected" not in res.stdout:
                    err_msg = res.stdout.strip() or res.stderr.strip()
                    add_log("FAIL", "PHP-SYNTAX", rel_path, "PHP Sözdizim Hatası", err_msg)
                else:
                    report["stats"]["pass"] += 1
            except Exception as e:
                add_log("FAIL", "PHP-EXEC-ERR", rel_path, str(e))

        # Node.js Kontrolü
        elif ext in {'.js', '.mjs', '.cjs'}:
            cmd = [NODE_EXE, "--check", file_path]
            try:
                res = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, timeout=10)
                if res.returncode != 0:
                    add_log("FAIL", "JS-SYNTAX", rel_path, "JS Sözdizim Hatası", res.stderr.strip())
                else:
                    report["stats"]["pass"] += 1
            except Exception as e:
                add_log("FAIL", "JS-EXEC-ERR", rel_path, str(e))

        # Python Kontrolü
        elif ext == '.py':
            cmd = [PYTHON_EXE, "-m", "py_compile", file_path]
            try:
                res = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, timeout=10)
                if res.returncode != 0:
                    add_log("FAIL", "PY-SYNTAX", rel_path, "Python Sözdizim Hatası", res.stderr.strip())
                else:
                    report["stats"]["pass"] += 1
            except Exception as e:
                add_log("FAIL", "PY-EXEC-ERR", rel_path, str(e))

        # JSON Kontrolü
        elif ext == '.json':
            try:
                json.loads(text_content)
                report["stats"]["pass"] += 1
            except Exception as e:
                add_log("FAIL", "JSON-PARSE", rel_path, f"Geçersiz JSON: {e}")

        # Kalıp Taraması (Güvenlik / Kalite)
        if ext in {'.php', '.js'}:
            for pat, cat, code, msg in BUG_PATTERNS:
                if pat.search(text_content):
                    if cat == "SECURITY":
                        if is_strict:
                            add_log("FAIL", code, rel_path, msg)
                        else:
                            add_log("WARN", code, rel_path, msg)
                    else:
                        add_log("INFO", code, rel_path, msg)

    if os.path.isfile(resolved_target):
        scan_file(resolved_target)
    else:
        for root, dirs, files in os.walk(resolved_target):
            dirs[:] = [d for d in dirs if d not in IGNORED_DIRS]
            for f in files:
                scan_file(os.path.join(root, f))

    if is_json:
        print(json.dumps(report, indent=2, ensure_ascii=False))
    else:
        print("=" * 60)
        print("🔍 HATA VE SÖZDİZİM KONTROL RAPORU (Python Sürümü)")
        print(f"Hedef  : {resolved_target}")
        print(f"Taranan: {report['scanned_files']} dosya")
        print("=" * 60)

        if not report["errors"] and not report["warnings"]:
            print("✅ TEBRİKLER: Hiçbir sözdizim veya kritik hata tespit edilmedi.")
        else:
            if report["errors"]:
                print(f"\n🔴 KRİTİK HATALAR ({len(report['errors'])} adet):")
                for idx, e in enumerate(report["errors"], 1):
                    print(f"  {idx}. [{e['code']}] {e['file']} → {e['message']}")
                    if e.get("detail"):
                        print(f"     Detay: {e['detail'].splitlines()[0]}")

            if report["warnings"]:
                print(f"\n🟡 UYARILAR ({len(report['warnings'])} adet):")
                for idx, w in enumerate(report["warnings"], 1):
                    print(f"  {idx}. [{w['code']}] {w['file']} → {w['message']}")

        print("\n" + "=" * 60)
        print("SONUÇ ÖZETİ:")
        print(f"  BAŞARILI (PASS) : {report['stats']['pass']}")
        print(f"  HATALI   (FAIL) : {report['stats']['fail']}")
        print(f"  UYARI    (WARN) : {report['stats']['warn']}")
        print(f"  BİLGİ    (INFO) : {report['stats']['info']}")
        print("=" * 60)

    sys.exit(1 if report["stats"]["fail"] > 0 else 0)

if __name__ == '__main__':
    run_checker()

File: mac/claude/claude-master-suite/scripts/error_checker.js
#!/usr/bin/env node
/**
 * error_checker.js — Kapsamlı Proje Hata ve Sözdizim Kontrol Aracı (Bug Hunter & QA)
 * 
 * Bu araç belirtilen dosya veya klasörü özyinelemeli olarak tarar:
 * 1. PHP Sözdizim Doğrulama (php -l)
 * 2. JavaScript / Node.js Sözdizim Doğrulama (node --check)
 * 3. Python Sözdizim Doğrulama (python3 -B -m py_compile)
 * 4. JSON Format ve Sözdizim Doğrulama (JSON.parse)
 * 5. UTF-8 BOM Tespiti (0xEF 0xBB 0xBF)
 * 6. Güvenlik ve Mantık Açığı Statik Analizi (eval, unescaped XSS, raw SQL, zayıf hash)
 * 7. Eksik require/include Tespiti
 * 
 * Kullanım:
 *   node error_checker.js [hedef_klasor_veya_dosya] [--fix-bom] [--json] [--strict]
 */

'use strict';

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

function getExecutor(canonicalPath, fallbackCmd) {
  if (canonicalPath && fs.existsSync(canonicalPath)) return `"${canonicalPath}"`;
  return fallbackCmd;
}

function resolvePython() {
  const candidates = [
    '/opt/homebrew/bin/python3',
    '/usr/local/bin/python3',
    '/usr/bin/python3'
  ];
  for (const c of candidates) {
    if (fs.existsSync(c)) return `"${c}"`;
  }
  try {
    execSync('python3 --version', { stdio: 'ignore' });
    return 'python3';
  } catch (e) {}
  return 'python';
}

const PHP_CMD = getExecutor('php', 'php');
const NODE_CMD = getExecutor('node', 'node');
const PYTHON_CMD = resolvePython();

// Argümanları ayrıştır
const args = process.argv.slice(2);
let targetPath = '.';
let fixBom = false;
let isJson = false;
let isStrict = false;

args.forEach(arg => {
  if (arg === '--fix-bom') fixBom = true;
  else if (arg === '--json') isJson = true;
  else if (arg === '--strict') isStrict = true;
  else if (!arg.startsWith('--')) targetPath = arg;
});

const resolvedTarget = path.resolve(targetPath);

if (!fs.existsSync(resolvedTarget)) {
  console.error(`❌ HATA: Hedef yol bulunamadı: ${resolvedTarget}`);
  process.exit(2);
}

const IGNORED_DIRS = new Set([
  'node_modules', 'vendor', '.git', 'storage', 'cache',
  'dist', 'build', '.vscode', '.idea', 'tmp', 'temp', '__pycache__'
]);

const report = {
  target: resolvedTarget,
  timestamp: new Date().toISOString(),
  scannedFiles: 0,
  stats: { pass: 0, fail: 0, warn: 0, info: 0 },
  errors: [],
  warnings: []
};

function addLog(type, code, file, message, detail = '') {
  if (type === 'FAIL') report.stats.fail++;
  else if (type === 'WARN') report.stats.warn++;
  else if (type === 'INFO') report.stats.info++;
  else if (type === 'PASS') report.stats.pass++;

  const entry = { type, code, file: path.relative(resolvedTarget, file) || path.basename(file), message, detail };
  if (type === 'FAIL') report.errors.push(entry);
  else if (type === 'WARN') report.warnings.push(entry);
}

// Dosya tarama fonksiyonu
function walk(dir) {
  const stat = fs.statSync(dir);
  if (!stat.isDirectory()) {
    checkFile(dir);
    return;
  }

  const items = fs.readdirSync(dir);
  for (const item of items) {
    if (IGNORED_DIRS.has(item)) continue;
    const fullPath = path.join(dir, item);
    try {
      const s = fs.statSync(fullPath);
      if (s.isDirectory()) {
        walk(fullPath);
      } else {
        checkFile(fullPath);
      }
    } catch (e) {}
  }
}

// Güvenlik ve mantık kalıp kontrolleri
const BUG_PATTERNS = [
  {
    regex: /eval\s*\(/i,
    category: 'SECURITY',
    code: 'SEC-EVAL',
    msg: 'eval() fonksiyonu kullanımı tespit edildi (uzaktan kod çalıştırma riski)'
  },
  {
    regex: /\bextract\s*\(\s*\$_(GET|POST|REQUEST)/i,
    category: 'SECURITY',
    code: 'SEC-EXTRACT',
    msg: 'extract($_GET/$_POST) ile kontrolsüz değişken enjeksiyonu riski'
  },
  {
    regex: /echo\s+\$_(GET|POST|REQUEST)\[[^\]]+\]\s*;/i,
    category: 'SECURITY',
    code: 'SEC-XSS',
    msg: 'Doğrudan kullanıcı girdisi echo ediliyor (XSS riski - htmlspecialchars zorunludur)'
  },
  {
    regex: /\$_(GET|POST|REQUEST)\[[^\]]+\]\s*\.\s*(['"][^'"]*SELECT|INSERT|UPDATE|DELETE)/i,
    category: 'SECURITY',
    code: 'SEC-SQLI',
    msg: 'Ham SQL sorgusu ile kullanıcı girdisi birleştiriliyor (SQL Enjeksiyonu riski)'
  },
  {
    regex: /\b(shell_exec|exec|system|passthru|popen|proc_open)\s*\(/i,
    category: 'SECURITY',
    code: 'SEC-CMD-EXEC',
    msg: 'Sistem komut yürütücü çağrısı tespit edildi'
  },
  {
    regex: /var_dump\s*\(|console\.log\s*\(/i,
    category: 'QUALITY',
    code: 'QUAL-DEBUG-LEAK',
    msg: 'Debug çıktısı (var_dump / console.log) canlı kodda kalmış olabilir'
  }
];

function checkFile(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  const allowedExts = ['.php', '.js', '.mjs', '.cjs', '.json', '.py', '.html', '.sql'];
  if (!allowedExts.includes(ext)) return;

  report.scannedFiles++;
  let buf;
  try {
    buf = fs.readFileSync(filePath);
  } catch (e) {
    addLog('FAIL', 'IO-ERR', filePath, 'Dosya okunamadı: ' + e.message);
    return;
  }

  // 1. BOM Kontrolü
  if (buf.length >= 3 && buf[0] === 0xEF && buf[1] === 0xBB && buf[2] === 0xBF) {
    if (fixBom) {
      fs.writeFileSync(filePath, buf.slice(3));
      addLog('WARN', 'BOM-FIXED', filePath, 'UTF-8 BOM tespit edildi ve otomatik onarıldı.');
      buf = buf.slice(3);
    } else {
      addLog('FAIL', 'BOM-DETECTED', filePath, 'Yasak UTF-8 BOM karakteri tespit edildi. (--fix-bom ile temizleyin)');
    }
  }

  const content = buf.toString('utf8');

  // 2. PHP Syntax Kontrolü (php -l)
  if (ext === '.php') {
    try {
      const out = execSync(`${PHP_CMD} -l "${filePath}"`, { stdio: ['pipe', 'pipe', 'pipe'] }).toString('utf8');
      if (!out.includes('No syntax errors detected')) {
        addLog('FAIL', 'PHP-SYNTAX', filePath, out.trim());
      } else {
        report.stats.pass++;
      }
    } catch (err) {
      const errOutput = (err.stderr ? err.stderr.toString('utf8') : '') || (err.stdout ? err.stdout.toString('utf8') : err.message);
      addLog('FAIL', 'PHP-SYNTAX', filePath, 'Sözdizim Hatası: ' + errOutput.split('\n')[0].trim(), errOutput.trim());
    }

    // Include / Require kontrolü
    const incRegex = /\b(require|require_once|include|include_once)\s*\(?\s*['"]([^'"]+\.php)['"]\s*\)?/g;
    let match;
    const fileDir = path.dirname(filePath);
    while ((match = incRegex.exec(content)) !== null) {
      const relTarget = match[2];
      if (!relTarget.startsWith('http://') && !relTarget.startsWith('https://')) {
        const potential = path.resolve(fileDir, relTarget);
        if (!fs.existsSync(potential) && !relTarget.includes('$')) {
          addLog('WARN', 'BROKEN-INCLUDE', filePath, `İçerilen dosya bulunamadı: ${relTarget}`);
        }
      }
    }
  }

  // 3. JavaScript / Node.js Syntax Kontrolü (node --check)
  if (['.js', '.mjs', '.cjs'].includes(ext)) {
    try {
      execSync(`${NODE_CMD} --check "${filePath}"`, { stdio: ['pipe', 'pipe', 'pipe'] });
      report.stats.pass++;
    } catch (err) {
      const errOutput = (err.stderr ? err.stderr.toString('utf8') : '') || err.message;
      addLog('FAIL', 'JS-SYNTAX', filePath, 'JavaScript Sözdizim Hatası', errOutput.trim());
    }
  }

  // 4. Python Syntax Kontrolü (python3 -B -m py_compile)
  if (ext === '.py') {
    try {
      execSync(`${PYTHON_CMD} -B -m py_compile "${filePath}"`, { stdio: ['pipe', 'pipe', 'pipe'] });
      report.stats.pass++;
    } catch (err) {
      const errOutput = (err.stderr ? err.stderr.toString('utf8') : '') || err.message;
      addLog('FAIL', 'PY-SYNTAX', filePath, 'Python Sözdizim Hatası', errOutput.trim());
    }
  }

  // 5. JSON Format Doğrulama
  if (ext === '.json') {
    try {
      JSON.parse(content);
      report.stats.pass++;
    } catch (err) {
      addLog('FAIL', 'JSON-PARSE', filePath, 'Geçersiz JSON formatı: ' + err.message);
    }
  }

  // 6. Güvenlik ve Kod Kalitesi Kalıp Taraması
  if (['.php', '.js'].includes(ext)) {
    BUG_PATTERNS.forEach(pat => {
      if (pat.regex.test(content)) {
        if (pat.category === 'SECURITY') {
          if (isStrict) {
            addLog('FAIL', pat.code, filePath, pat.msg);
          } else {
            addLog('WARN', pat.code, filePath, pat.msg);
          }
        } else {
          addLog('INFO', pat.code, filePath, pat.msg);
        }
      }
    });
  }
}

// Ana yürütme
walk(resolvedTarget);

// Çıktı Üretimi
if (isJson) {
  console.log(JSON.stringify(report, null, 2));
} else {
  console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
  console.log(`🔍 HATA VE SÖZDİZİM KONTROL RAPORU (Bug Hunter & QA)`);
  console.log(`Hedef  : ${resolvedTarget}`);
  console.log(`Zaman  : ${report.timestamp}`);
  console.log(`Taranan: ${report.scannedFiles} dosya`);
  console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');

  if (report.errors.length === 0 && report.warnings.length === 0) {
    console.log('✅ TEBRİKLER: Hiçbir sözdizim veya kritik hata tespit edilmedi.');
  } else {
    if (report.errors.length > 0) {
      console.log(`\n🔴 KRİTİK HATALAR (${report.errors.length} adet):`);
      report.errors.forEach((e, idx) => {
        console.log(`  ${idx + 1}. [${e.code}] ${e.file}`);
        console.log(`     Mesaj: ${e.message}`);
        if (e.detail) console.log(`     Detay: ${e.detail.split('\n')[0]}`);
      });
    }

    if (report.warnings.length > 0) {
      console.log(`\n🟡 UYARILAR & GÜVENLİK İPUÇLARI (${report.warnings.length} adet):`);
      report.warnings.forEach((w, idx) => {
        console.log(`  ${idx + 1}. [${w.code}] ${w.file} → ${w.message}`);
      });
    }
  }

  console.log('\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
  console.log(`SONUÇ ÖZETİ:`);
  console.log(`  BAŞARILI (PASS) : ${report.stats.pass}`);
  console.log(`  HATALI   (FAIL) : ${report.stats.fail}`);
  console.log(`  UYARI    (WARN) : ${report.stats.warn}`);
  console.log(`  BİLGİ    (INFO) : ${report.stats.info}`);
  console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
}

// Çıkış Kodu
if (report.stats.fail > 0) {
  process.exit(1);
} else {
  process.exit(0);
}
